### DISCUSSION TITLE
SoftConsole Load ELF File to DDR for Debugging

### DISCUSSION THREAD

**Discussion by MartinDKing** [polarfire-soc/polarfire-soc-documentation#332] [Bare metal embedded software] [answered]
I'm using the FPSoC Icicle Dev Kit programmed with MPFS_ICICLE_KIT_BASE_DESIGN_2023_02

I have been hacking the mpfs-rpmsg-freertos project from the polarfire-soc-amp-examples.

Specifically I have:

1. Disabled REMOTEPROC in the rules.mk file - I didn't want to fire up Linux to debug this.
2. Added a couple of extra items to the menu to read / write stuff to FIC3
3. Removed the -nostdlib and -nodefaultlibs from the Makefile - I needed the some extra clib bits & pieces
4. Created a yaml file for the packager that only has the mpfs-rpmsg-remote-elf (exec-addr: '0x91C00000', owner-hart: u54_4, priv-mode: prv_m, skip-opensbi: true}

Currently I:

1. Build in SoftConsole
2. Package the ELF
3. Write to SD card
4. Boot the Icicle Kit from SD (my code runs)
5. Attach the debugger in SoftConsole using mpfs-rpmsg-freertos hw one-hart attach.remote
6. Debug the code..

Finally, the question:

I expected to be able to attach the debugger and for SoftConsole to be able to load the ELF into DDR and debug without having to write it to SD (or eNVM).

Without going through the myriad of options that I have tried that failed in SoftConsole settings, please can someone either point me to the documentation, give me the settings required to make this work or put me out of my misery and tell my why this cannot work?

Thanks!

Kind regards,
Martin.

**Comment by MartinDKing** [comment]
I've added 'set remotetimeout 7' to the Edit Configuration->Debugger->GDB Client Setup->Commands field as suggested [here](https://mi-v-ecosystem.github.io/SoftConsole-Documentation/SoftConsole-v2022.2-RISC-V/using_softconsole/debugging.html)
Also with:
1. Ticked the Edit Configuration->Startup->Initial Reset (left the type as Init)
2. Ticked the Edit Configuration->Startup->Load Executable
3. Not Ticked the Edit Configuration->Startup->Debug in RAM
4. Not Ticked the Edit Configuration->Startup->Pre-run/Restart reset

Then, although it looks like it's failing with some error messages about Hart 4 not coming out of reset in 30s and a bunch of warnings about unexpected negative acknowledgements, after around 60s, it did load the code and stop in the pre-main start.

As suggested in the Console window, I have added 'riscv set_reset_timeout_sec' to the board/microsemi-riscv.cfg file. I tried doubling the default to 60, but with the same outcome albeit having to wait an extra 30s.

So, I have 'riscv set_reset_timeout_sec 1' in the cfg file, and it still works / gives error & warning messages, but eventually loads the code.

There's a slight caveat, when it finishes loading the code, it looks like it has halted somewhere in the code, but a single step shows it's really in the pre-main startup code.

Is this as good as debugging gets on the PFSoC?

**Reply by vfalanis** [reply, marked as answer]
Hi @MartinDKing ,

Is there a specific reason why you disabled remoteproc in the mpfs-rpmsg-freertos project? If using AMP with rpmsg communication, remoteproc must be enabled.

Have you tried the following?

1. Stop the current amp application from the bare metal/RTOS context using the following command
```
echo stop >/sys/class/remoteproc/remoteproc0/state
```
2. Build the mpfs-rpmsg-freertos application in SoftConsole. You will need to build using the REMOTEPROC=1 flag

![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/5502469/6f5bcafc-30ba-4c02-8fe6-481803a93fcf)

3. Transfer the elf to Linux using something like FTP, SSH, wget. The elf file should be placed in the `/sys/class/remoteproc/remoteproc0/firmware` directory and should be called `rproc-miv-rproc-fw`

4. Start  the new freertos application using the following command:
```
echo start >/sys/class/remoteproc/remoteproc0/state
```
5. Use the `mpfs-rpmsg-freertos hw one-hart attach remote.launch` to attach to the GDB session.

By using this method, you can debug faster with remoteproc and without the need of generating a new HSS payload each time.

Hope this helps.
